rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,394 @@
+repository, with TensorFlow as a dependency.
+
+## Motivation
+
+### Build times
+
+Building the open-source TensorFlow project end-to-end is an extensive exercise. 
+With a standard GCP instance, it might take more than one hour to finish the whole
+build process (it might take longer with a Mac laptop). Although the local build 
+cache might help speed up the follow-up builds, the initial time cost is too 
+high for regular software development workflows. Internally, Google has a
+distributed build and caching service, which Googlers heavily rely on,
+that can build TensorFlow and run all Keras tests within 5 mins. Sadly,
+we can't expose this to external contributors.
+
+Currently, any contribution to Keras code will require building all of
+TensorFlow, which is quite expensive to do for average users.
+Having a separate repository will allow the Keras package to be built
+without building TensorFlow. This should greatly improve the 
+velocity of open-source developers when they contribute to Keras code.
+
+### Community Benefit
+
+The difficulty of building TensorFlow from scratch in order to make a PR
+to Keras code has been a significant source of issues:
+
+* It discouraged contributions, since many external developers couldn't test
+their changes and make sure they were correct.
+* External developers would send unverified PRs, and Google reviewers spend time back 
+and forth, fixing the PR. Sometimes PR is just not moving forward because of the
+lengthy feedback loop.
+
+With the new standalone Keras repository, external contributors
+should experience much shorter turn-around time when 
+building/testing Keras, since they don't need to build TensorFlow anymore.
+This should  have a positive impact on building a vibrant open-source
+developer community.
+
+In addition, by getting the Keras team at Google to start developing Keras
+using the same public tools and infrastructure as third-party developers,
+we make the development process more transparent and more community-oriented.
+
+### TensorFlow API modularity
+
+There are other side-benefits if we split the repository. Currently, Keras
+has to rely on a number of private TensorFlow APIs. However, a litmus test
+of the quality of the public TensorFlow low-level APIs is that they should
+be strictly sufficient to a higher-level API like Keras.
+After splitting the repository, Keras will have to import TensorFlow and 
+rely exclusively on public APIs. If Keras still ends up using TensorFlow
+private features, it  might be an indication of tight coupling of
+implementation details. If certain private features are extensively used,
+we might want to consider exposing them  as public low level API.
+
+This design is also aligned with the design for
+[Modular TensorFlow](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md), 
+which splits the TensorFlow project into smaller components that are not
+tightly coupled together.
+
+
```

Comment:
There are also other advantaged worth mentioning. For example a keras repository makes it easier to share some of the management with the keras SIG than the single shared TF repository.

We need to also consider drawbacks of this decision here.

Specifically:
 1. cross-repository changes will be more expensive to submit (and now most keras changes would cross the repository boundary)
 2. CI will be more complicated (we'll need to test a product of TF and keras versions as opposed to a single shared version). This is complicated further by pulling these two repositories inside google's version control system, for some of us.
 3. Previous experience splitting tf.estimator with the same type of argument justifying it has been considered a failure. How are we sure keras won't fall prey to the same issues?

Reply:
Specifically, I'd like to see this proposal cover alternative approaches to the problems we've outlined, including a single repository with independent continuous integration, a tf-python repository separate from a tf-core repository, etc.

Reply:
Chatted w/ @alextp a bit about this,

The following are all really important goals:

1. Make it easy for the community to contribute to & manage Keras (without having to build all the rest of TF to do so)
2. Layer the external TF API generation so Keras can be transitioned to using entirely public TF symbols (with actual tf.* calls)
3. Enable building a keras-only pip package separate from the standard tf package, which users will be able to use via just `import Keras`

It would be useful to explore possible solutions that don't create the same degree of development / release syncing overhead.

E.g. in addition to @alextp's suggestions above, we could explore some sort of transitional plan that partially or mostly accomplishes the above 3 goals while keeping Keras and TF in the same repo w/ a synced release cycle. 
Once all the circular dependencies have truly been fixed and Keras already uses the public TF APIs directly, we could revisit splitting Keras into a fully separate repo w/ its own release cycle and source of truth.

Reply:
+1 
the limited experience with tf.estimator suggests that we should consider the drawbacks seriously and convince ourselves that they are something we would be OK dealing with on a day to day basis (or consider alternatives). For e.g. the 2 stage submitting process (with 1 have google as source of truth and one having github as source of truth) can be onerous and error prone (accounting for versions, rollbacks etc)

Reply:
Done. Thanks for the suggestion. I have added a section for CI and alternative approaches. we should discuss the alternatives in the design review.